extensions/amp-ad/0.1/amp-ad.js

```diff
@@ -65,21 +44,26 @@ export class AmpAd extends AMP.BaseElement {
     }
     // TODO(tdrl): Check amp-ad registry to see if they have this already.
     if (!a4aRegistry[type] ||
-        !a4aRegistry[type](this.win, this.element)) {
+        !a4aRegistry[type](this.getWin(), this.element)) {
       // Network either has not provided any A4A implementation or the
       // implementation exists, but has explicitly chosen not to handle this
       // tag as A4A.  Fall back to the 3p implementation.
       return new AmpAd3PImpl(this.element);
     }
-    // TODO(dvoytenko): Reimplement a4a via `upgradeCallback`. It will look
-    // as following:
-    /*
-      const extensions = extensionsFor(this.win);
-      return extensions.loadElementClass(extensionTag).then(ctor => {
+    const extensionTagName = networkImplementationTag(type);
+    this.element.setAttribute('data-a4a-upgrade-type', extensionTagName);
+    return extensionsFor(this.getWin()).loadElementClass(extensionTagName)
```

Comment:
I've always been confused what file provides the `amp-ad-network-${type}-impl` element class...